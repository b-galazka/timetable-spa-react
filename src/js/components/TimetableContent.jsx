import React, { Component } from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import Table from './Table';

import texts from '../../json/texts';

import '../../scss/timetableContent.scss';
import '../../scss/animations.scss';

function mapStateToProps(state) {

    const { timetableObject } = state;

    return {
        lastModified: timetableObject.lastUpdateDate
    };
}

class TimetableContent extends Component {

    constructor() {

        super();

        this.state = {
            animationClass: true
        };

        this.removeAnimationClass = this.removeAnimationClass.bind(this);
    }

    render() {

        return (
            <section className="timetable__wrapper"
                className={

                    classNames({
                        'timetable__wrapper': true,
                        animated: this.state.animationClass
                    })
                }

                onAnimationEnd={this.removeAnimationClass}
            >
                <div className="timetable__content">
                    <Table />
                </div>

                <footer className="timetable__footer">
                    <p>{texts.lastModified}: {this.convertDate()}</p>
                </footer>
            </section>
        );
    }

    convertDate() {

        const date = new Date(this.props.lastModified);

        const year = date.getFullYear();
        const day = this.formatTimeUnit(date.getDate());
        const month = this.formatTimeUnit(date.getMonth() + 1);
        const hour = this.formatTimeUnit(date.getHours());
        const minutes = this.formatTimeUnit(date.getMinutes());

        return `${day}/${month}/${year} ${hour}:${minutes}`;
    }

    formatTimeUnit(unit) {

        return (unit < 10) ? `0${unit}` : unit;
    }

    removeAnimationClass() {

        this.setState({
            animationClass: false
        });
    }
}

TimetableContent.propTypes = {

    // redux
    lastModified(props, propName) {

        const propValue = props[propName];

        if (new Date(propValue).toString() === 'Invalid Date') {

            return new Error('props.lastModified validation error');
        }
    }
};

export default connect(mapStateToProps)(TimetableContent);