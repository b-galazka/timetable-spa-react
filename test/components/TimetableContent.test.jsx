import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import renderer from 'react-test-renderer';

import TimetableContent from '../../src/js/components/TimetableContent';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('../../src/js/components/Table', () => 'Table');
jest.mock('../../src/json/texts', () => ({ lastModified: 'last modified' }));

describe('TimetableContent component', () => {

    it('should render with date in valid format', () => {

        const tree = renderer.create(
            <TimetableContent.WrappedComponent
                lastModified="Thu Dec 20 2018 10:47:46 GMT+0100"
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with date in DD/MM/YYYY HH:MM format', () => {

        const tree = renderer.create(
            <TimetableContent.WrappedComponent
                lastModified="Sat Jun 02 2018 09:05:46 GMT+0200"
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should remove "animated" CSS class when animation is finished', () => {

        const content = mount(
            <TimetableContent.WrappedComponent
                lastModified="Thu Dec 20 2018 10:47:46 GMT+0100"
            />
        );

        content.find('.timetable__wrapper').simulate('animationEnd');

        expect(content.find('.timetable__wrapper').hasClass('animated')).toBe(false);
    });
});