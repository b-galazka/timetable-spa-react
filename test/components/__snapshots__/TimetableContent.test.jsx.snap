// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimetableContent component should render with date in DD/MM/YYYY HH:MM format 1`] = `
<section
  className="timetable__wrapper animated"
  onAnimationEnd={[Function]}
>
  <div
    className="timetable__content"
  >
    <Table />
  </div>
  <footer
    className="timetable__footer"
  >
    <p>
      last modified
      : 
      02/06/2018 09:05
    </p>
  </footer>
</section>
`;

exports[`TimetableContent component should render with date in valid format 1`] = `
<section
  className="timetable__wrapper animated"
  onAnimationEnd={[Function]}
>
  <div
    className="timetable__content"
  >
    <Table />
  </div>
  <footer
    className="timetable__footer"
  >
    <p>
      last modified
      : 
      20/12/2018 10:47
    </p>
  </footer>
</section>
`;
